### ISSUE TITLE
Passthrough USB Device

### ISSUE THREAD

**Issue by Gothicaria** [dockur/windows#140]
Hi there, I want to be able to passthrough my USB devices like you can in a VM, in particular my iCue Link Hub.

How exactly do I do this?

![image](https://github.com/dockur/windows/assets/158237482/51f0cbf9-266d-4343-84b2-1a93cd41fe58)

Thankyou in advance!
Erin

**Comment by Gothicaria** [comment]
I added the variable and the extra parameter to the docker container, (was 001/006) in my case. iCue hub not being detected by iCue in the WID so I am assuming it was not successful. :(

**Comment by Gothicaria** [comment]
Still no dice. Also, I do apologise my skills are only intermediate when it comes to unraid/dockers. But I do have a good understanding of what you are asking me to execute, I will let you know if I do not understand :)

**Comment by Gothicaria** [comment]
So this is what I get: 
![image](https://github.com/dockur/windows/assets/158237482/7849bf61-202e-4602-bd3e-4ade719f9cf7)

**Comment by Gothicaria** [comment]
> Run `sudo apt-get install usbutils`

From inside the shell access or just from a new terminal?

**Comment by Gothicaria** [comment]
> From inside the shell of the container (where you got the command not found error).

replied with 
bash: sudo: command not found

**Comment by Gothicaria** [comment]
![image](https://github.com/dockur/windows/assets/158237482/6140ea32-c3c4-468d-9e3d-bad3a9adce07)

**Comment by Gothicaria** [comment]
That seemed to work:
![image](https://github.com/dockur/windows/assets/158237482/e4d5e08a-5382-498b-8665-b7d3ddcc98e6)

**Comment by Gothicaria** [comment]
![image](https://github.com/dockur/windows/assets/158237482/8e04310e-4581-4773-b720-862b01eb2ad1)

**Comment by Gothicaria** [comment]
Ok so im certain I followed this correctly
![image](https://github.com/dockur/windows/assets/158237482/64d6af36-1e09-4dcd-8af6-d1eede6f3163)

![image](https://github.com/dockur/windows/assets/158237482/57372f89-2cfa-43ac-9275-1434ee935532)

So the container is obviously seeing all of the USB devices, but not passing them through to the VM?

**Comment by Gothicaria** [comment]
Also not running in privliaged

**Comment by Gothicaria** [comment]
> -device djskskbendjsjsmnw

First one, no change

Second nonsense argument and the docker still started

**Comment by Gothicaria** [comment]
> docker exec -it containerid /bin/bash

when I try the 1susb command again to check to see what it is seeing its gone back to the original error

**Comment by Gothicaria** [comment]
> lsusb

Sorry ignore that, it os seeing it
I made a typo

**Comment by Gothicaria** [comment]
> Maybe the problem is that you gave it the Name ARGUMENTS but the Key should be ARGUMENTS

OMG I love you that was it!! Changed key to Arguments and used -device usb-host,hostdevice=/dev/bus/usb/001/006

![image](https://github.com/dockur/windows/assets/158237482/04f64925-da71-43e2-ac6c-d39b9a9ba9bd)

![image](https://github.com/dockur/windows/assets/158237482/893393f9-3835-47c7-b687-68f886311aac)
 I cant thank you enough!!!!

**Comment by Gothicaria** [comment]
Thank you again, you are a gem. Hopefully this will help others the same. Have a wonderful day!

**Comment by Gothicaria** [comment]
Just an update for you. It appears at some point the pass-through stopped working and I noted that the iCue link hub changed from Bus001/Device6 to Bus001/Device5. Going back through the steps you gave me I was able to fix it but it is likely going to happen again. Is there a way around this and/or why do you think it would change like that? I had not made any changes to the hardware etc.

**Comment by Gothicaria** [comment]
Yeah its weird as I didnt touch it. I will have to try the way you mentioned in the link and see what happens. Thankyou!

**Comment by teije** [comment]
I have attempted to passthrough a USB connected ESP32. 

**TL:DR;**
As far as I understand I have taken all steps already described in this thread, but no luck so far. 
Does anyone have any tips on how to get this to work?


**Steps taken so far:**

I have identified it to be device on: `Bus003 Device005` with the `vendorid:productid` being:  `1a86:55d4`

1. When I run `lsusb` on the Unraid host machine I get:
```
  Bus 004 Device 002: ID 090c:1000 Silicon Motion, Inc. - Taiwan (formerly Feiya Technology Corp.) Flash Drive
  Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
> Bus 003 Device 005: ID 1a86:55d4 QinHeng Electronics USB Single Serial
  Bus 003 Device 004: ID 051d:0002 American Power Conversion Uninterruptible Power Supply
  Bus 003 Device 003: ID 1a40:0101 Terminus Technology Inc. Hub
  Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
  Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
  Bus 001 Device 004: ID 8087:0029 Intel Corp. AX200 Bluetooth
  Bus 001 Device 002: ID 048d:8297 Integrated Technology Express, Inc. IT8297 RGB LED Controller
  Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

2. When I run `lsusb` on the 'WindowsinDocker'-container I get:
```
  Bus 001 Device 001: ID 1d6b:0002 Linux 6.1.74-Unraid xhci-hcd xHCI Host Controller
  Bus 001 Device 002: ID 048d:8297 ITE Tech. Inc. ITE Device(8595)
  Bus 001 Device 004: ID 8087:0029  
  Bus 002 Device 001: ID 1d6b:0003 Linux 6.1.74-Unraid xhci-hcd xHCI Host Controller
  Bus 003 Device 001: ID 1d6b:0002 Linux 6.1.74-Unraid xhci-hcd xHCI Host Controller
  Bus 003 Device 003: ID 1a40:0101  USB 2.0 Hub
  Bus 003 Device 004: ID 051d:0002 American Power Conversion Back-UPS BX950MI  FW:295202G -302202G 
> Bus 003 Device 005: ID 1a86:55d4  USB Single Serial
  Bus 004 Device 001: ID 1d6b:0003 Linux 6.1.74-Unraid xhci-hcd xHCI Host Controller
  Bus 004 Device 002: ID 090c:1000 Samsung Flash Drive FIT 
```

3. I have attempted to use the /dev/bus/usb/bus-id/device-id ARGUMENTS configuration:
`-device usb-host,hostdevice=/dev/bus/usb/003/005`
![afbeelding](https://github.com/dockur/windows/assets/33690639/ac1ced81-9b70-4b97-afb0-57f134e53b18)
    - This configfuration prevents the container from starting
      `❯ ERROR: qemu-system-x86_64: -device usb-host,hostdevice=/dev/bus/usb/003/005: failed to open /dev/bus/usb/003/005: No such file or directory`
    - Screenshot:
       ![afbeelding](https://github.com/dockur/windows/assets/33690639/bf5866a1-48d1-494b-baa4-920463c0f19d)


4. I have attempted to use the vendorid:productid ARGUMENTS configuration:
`-device usb-host,vendorid=0x1a86,productid=0x55d4`
![afbeelding](https://github.com/dockur/windows/assets/33690639/bda5b3f9-d077-4995-862b-a86947e76ec8)
    - This configuration allows the container to start, which seemed hopeful at first, but I don't think the device is actually properly passedthrough.     
    - Below is a screenshot of what I can see in USBDeview installed on the Windows11 installation inside the docker container
    - I am unsure of what the Port_#0005.Hub_#0001-device is, but I do not think that it is the ESP32.
    ```
    Port_#0005.Hub_#0001	USB Input Device	HID (Human Interface Device)	Yes	Yes	No	No		28754-0000:00:02.0-1	3/13/2024 8:00:07 PM	3/12/2024 9:19:39 AM	0627	0001	0.00		03	00	00		DOCKERW-J5V62GK			6&25ad3554&0	HidUsb	@input.inf,%HID.SvcDesc%;Microsoft HID Class Driver	hidusb.sys		(Standard system devices)		100 mA	2.00	USB Input Device	10.0.22621.2506	HID_Inst.NT	input.inf	USB\VID_0627&PID_0001\28754-0000:00:02.0-1	Removable, UniqueID, SurpriseRemovalOK	3/12/2024 9:19:39 AM	3/12/2024 9:19:39 AM	3/13/2024 8:00:07 PM	
    ``` 
    ![afbeelding](https://github.com/dockur/windows/assets/33690639/1a5fd8a8-f81a-415b-a63f-c530f62ecbab)

**Comment by K-J-VV** [comment]
@kroese sorry to revive a previously closed problem. Thanks to you & @Gothicaria I managed to get some of my iCUE devices to pass though but I'm having trouble with one in particular. Not sure if you have any advice on how to get this particular device to be detected/used by the container?  

Here is the error message: 

`❯ Booting Windows using QEMU v9.2.2...
❯ ERROR: qemu-system-x86_64: -device usb-host,hostdevice=/dev/bus/usb/005/005: Warning: speed mismatch trying to attach usb device "CORSAIR iCUE LINK AIO LCD Scree" (high speed) to bus "xhci.0", port "4.1" (full speed)
qemu-system-x86_64: -device usb-host,hostdevice=/dev/bus/usb/005/005: failed to open host usb device /dev/bus/usb/005/005
`

**Comment by K-J-VV** [comment]
> [@K-J-VV](https://github.com/K-J-VV) This should now be fixed in v4.27, can you please try it out and let me know if it makes a difference?

Just upgraded and receiving the following boot error 

` Booting Windows using QEMU v9.2.2...
❯ ERROR: qemu-system-x86_64: Serial: Could not open 'Serial': No such file or directory`
